export/export_pipeline/compute_collision_representation.py

```diff
@@ -0,0 +1,427 @@
+                elif (o.type == 'EMPTY') and not o.children:  # PhysicsConstraintComponent (i.e. joint)
+                    joint_name = utils.name_in_names(name, joints_info.keys())
+                    components_info['joints'][joint_name] = {
+                        'pos': xyz,
+                        'quat': quat,
+                    }
+
+            filename = osp.join(obj_dir, 'components.json')
+            with open(filename, 'w') as f:
+                json.dump(components_info, f, indent=4)
+
+            # group the leaves by group_id
+            # TODO(samarth) group name 
+            geom_groups = defaultdict(list)
+            for geom in geoms:
+                geom_groups[geom[-1]].append(geom[:-1])
+            
+            for geom_group in geom_groups.values():
+                geom_dir, geom, decompose_method = geom_group[0]
+                leaf_name = geom.name.split('.')[0]
+                
+                # merge the geoms
+                for l in geom_group[1:]:
+                    l[1].select_set(True)
+                    shutil.rmtree(l[0])
+                geom.select_set(True)
+                bpy.context.view_layer.objects.active = geom
+                bpy.ops.object.join()
+                
+                # Calculate the number of faces in the mesh
+                poly_count = len(geom.data.polygons)
+                print("Polygon count: ", poly_count)
+                
+                # Get the adaptive decimation ratio
+                decimation_ratio = decimation_lookup(poly_count)
+                    
+                # Add the Decimate modifier to the selected object
+                print("Applying decimation with a ratio: ", decimation_ratio)
+                decimate_modifier = geom.modifiers.new(name='Decimate', type='DECIMATE')
+                decimate_modifier.ratio = decimation_ratio
+
+                # Exporting in stl.
+                stl_filepath = osp.join(geom_dir, f"{leaf_name}.stl")
+                if (not self.rerun) or (not osp.isfile(stl_filepath)):
+                    print("Exporting STL file...")
+                    bpy.ops.export_mesh.stl(
+                        filepath=stl_filepath,
+                        check_existing=False,
+                        use_selection=True,
+                        use_mesh_modifiers=True,
+                        batch_mode='OFF',
+                        global_scale=1.0,
+                        axis_forward='Y',
+                        axis_up='Z',
+                        ascii=False,
+                    )
+                bpy.ops.object.select_all(action='DESELECT')
+
+                params = copy.deepcopy(self.p)
+                params['collision_representation']['acd']['decompose_method'] = decompose_method
+                process_mesh_args.append((stl_filepath, params, self.rerun))
+
+        print("Convex solid decomposition...")
+        i = 0
+        while True:
```

Comment:
This while loop is ugly. Why not have a for-loop since you already know the number of iterates when the loop is executed. Additionally, this code would reduce CPU utilization in the case where some iterates are slow and some are fast. It seems like this could easily happen because some COACD invocations could be really complicated, and some are trivial.

You'd rather specify the full workload when calling `map`, to maximize the opportunities of each worker to steal other work if it finishes its own work quickly. Does `map` not allow you to do this?

Finally, any pipeline stage that uses `mp` to provide a single-process reference implementation that doesn't use `mp`.

Reply:
Full workload consumes a large amount of memory. Please see the explanation I have written as a comment in the latest version. I have also replaced the while loop with a more succinct for loop.